ts/client/src/accounts/mangoAccount.ts

```diff
@@ -1628,6 +1676,243 @@ export class PerpPosition {
+    let wbpn; // B
+    let qpn = this.quotePositionNative.sub(
+      this.getUnsettledFunding(perpMarket),
+    );
+    {
+      const baseNative = this.getBasePositionNative(perpMarket);
+      if (baseNative.isZero()) {
+        return MINUS_ONE_I80F48();
+      }
+
+      if (baseNative.isNeg()) {
+        const ret = PerpPosition.getAggregatePerpOrderSizeAndQuoteNative(
+          perpMarket,
+          sortedOo,
+          PerpOrderSide.ask,
+          tillPriceExclusive,
+        );
+        wbpn = perpMarket.maintBaseLiabWeight.mul(
+          baseNative.sub(
+            I80F48.fromI64(ret.sizeLots).mul(
+              I80F48.fromI64(perpMarket.baseLotSize),
+            ),
+          ),
+        );
+        qpn = qpn.add(ret.quoteNative);
+      } else {
+        const ret = PerpPosition.getAggregatePerpOrderSizeAndQuoteNative(
+          perpMarket,
+          sortedOo,
+          PerpOrderSide.ask,
+          tillPriceExclusive,
+        );
+        wbpn = perpMarket.maintBaseAssetWeight.mul(
+          baseNative.add(
+            I80F48.fromI64(ret.sizeLots).mul(
+              I80F48.fromI64(perpMarket.baseLotSize),
+            ),
+          ),
+        );
+        qpn = qpn.sub(ret.quoteNative);
+      }
+    }
+
+    if (wbpn.isZero()) {
+      return MINUS_ONE_I80F48();
+    }
+
+    // Health contribution of perp position in question would be considered separately
+    const mAClone: MangoAccount = cloneDeep(mangoAccount);
+    const ppToBeErased: PerpPosition = mAClone.getPerpPosition(
+      perpMarket.perpMarketIndex,
+    )!;
+    ppToBeErased.marketIndex =
+      PerpPosition.PerpMarketIndexUnset as PerpMarketIndex;
+    const healthCacheWoPerpPosition = HealthCache.fromMangoAccount(
+      group,
+      mAClone,
+    );
+
+    // Unused, TODO what to do with this?
+    // const basePrice = qpn.neg().div(wbpn);
+
+    const healthContributionWithoutSettleToken =
+      healthCacheWoPerpPosition.healthExcludingToken(
+        HealthType.maint,
+        perpMarket.settleTokenIndex,
+      ); // C
+
+    const settleTokenWeight = healthContributionWithoutSettleToken.lt(
+      ZERO_I80F48(),
+    )
+      ? settleTokenBank.maintAssetWeight
+      : settleTokenBank.maintLiabWeight;
+    const settleTokenSpot =
+      healthCacheWoPerpPosition.tokenInfos[
+        healthCacheWoPerpPosition.getOrCreateTokenInfoIndex(settleTokenBank)
+      ].balanceSpot;
+    const otherPerpsSettledInSettleTokenUhuPnl =
+      healthCacheWoPerpPosition.perpInfos
+        .filter((pi) => pi.settleTokenIndex == perpMarket.settleTokenIndex)
+        .reduce(
+          (sum, pi) => sum.add(pi.healthUnsettledPnl(HealthType.maint)),
+          ZERO_I80F48(),
+        );
+    const openBookFree = mangoAccount.getSerum3TokenFreeBalance(
+      group,
+      settleTokenBank,
+    );
+    const settleTokenContribution = settleTokenSpot
+      .add(otherPerpsSettledInSettleTokenUhuPnl)
+      .add(openBookFree)
+      .add(qpn); // T
```

Comment:
It might be preferable to build C and T by subtraction, instead of constructively.

```
C = maintHealth 
    - tokenContribution(effectiveTokenPos[settleToken])
T = tokenContribution(effectiveTokenPos[settleToken]) 
    - perpPos.hupnl 
    + quote pos native
```

That way this code doesn't need to know as much about how health is constructed, just about perp health.